### PR TITLE
Support multiple compress stream types with same topology

### PR DESCRIPTION
**Pull request by dbaluta**
With this changes we can play AAC/MP3 streams with the same topology. This brings the advantage that we do not need to reboot/or change topology in order to playd diffferent stream types.

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
@dbaluta Good to hear this is working, do we need any kernel updates for this to work ?
@plbossart fyi.

**Comment by dbaluta** [comment]
@lgirdwood no need for any kernel updates.

**Comment by lgirdwood** [comment]
@dbaluta do you have an example of the command line that would be used in this use case ? I'm guessing it's just 
```
cplay file.mp3
cplay file.aac
```
i.e. there is no kcontrol to switch between aac and mp3 ?

File: src/audio/module_adapter/module/cadence.c [hunk 1]
```diff
@@ -35,6 +35,24 @@ enum cadence_api_id {
 
 #define DEFAULT_CODEC_ID CADENCE_CODEC_WRAPPER_ID
 
+/* we divide Cadence param id space as follows:
+ * - 0x00000000 - 0x7FFFFFFF, Cadence library ids which come
+ *   from xa_*_dec_api.h
+ * - 0x80000000 - 0xFFFFFFFF, SOF remapping of Cadence library ids
+ */
```
Comment (plbossart) [inline src/audio/module_adapter/module/cadence.c:42]: Unable to follow the commit message

"
We will do this in two steps:

1) Allow parameters to be specified in topology.
2) Send parameters via snd_codec structure.

This patch addresses step 1).

We need both steps, because some parameters like `pcm width size` for
compress streams is not stored in snd_codec and cannot be retrieved
from the actual compress file.
"

Not following at all which parameters need to be stored in the topology?

File: tools/topology/topology1/development/sof-imx8-compr-wm8960.m4 [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+include(`platform/imx/imx8.m4')
+
+
+# Post process setup config
+
+ #codec Post Process setup config
+#
+# Define the pipelines
+#
+# PCM0 <----> volume <-----> SAI3 (wm8960)
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply tools/topology/topology1/development/sof-imx8-compr-wm8960.m4:29]: playback only, copy paste error.

File: tools/topology/topology1/development/sof-imx8-compr-wm8960.m4 [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+DAI_ADD(sof/pipe-dai-playback.m4,
+	1, SAI, 1, sai1-wm8960-hifi,
+	PIPELINE_SOURCE_1, 2, s32le,
+	1000, 0, 0, SCHEDULE_TIME_DOMAIN_DMA)
+
+
+# PCM Low Latency, id 0
+
+dnl PCM_DUPLEX_ADD(name, pcm_id, playback, capture)
+COMPR_PLAYBACK_ADD(Port0, 0, PIPELINE_PCM_1)
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply tools/topology/topology1/development/sof-imx8-compr-wm8960.m4:88]: only playback, will fix the comments as they are copy pasted.

File: tools/topology/topology1/development/sof-imx8mp-compr-aac-wm8960.m4 [hunk 4]
```diff
@@ -38,12 +38,14 @@ define(`CA_UUID', cadence_codec_uuid)
...
 `	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'
 `	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,'
-`	0x07,0x00,0x00,0x00,0x0C,0x00,0x00,0x00,'
-`	0x18,0x00,0x00,0x00,0x03,0x00,0x00,0x00,'
-`	0x0C,0x00,0x00,0x00,0x02,0x00,0x00,0x00"''
+`	0x07,0x20,0x00,0x80,0x0C,0x00,0x00,0x00,'
+`	0x18,0x00,0x00,0x00,0x03,0x20,0x00,0x80,'
+`	0x0C,0x00,0x00,0x00,0x02,0x00,0x00,0x00,'
+`	0x00,0x10,0x00,0x80,0x0C,0x00,0x00,0x00,'
+`	0x18,0x00,0x00,0x00"''
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply tools/topology/topology1/development/sof-imx8mp-compr-aac-wm8960.m4:48]: will add some comments. Indeed this is hard to track, but this is the way parameters are designed to be sent to firmware. My plan is to reduce config parameters kept in topology.

I think I can reduce it to just 1: PCM width size. PCM width size is not a parameter of the compress file but just a 'design' decision of the pipeline.

File: src/audio/module_adapter/module/cadence.c [hunk 5]
```diff
@@ -35,6 +35,24 @@ enum cadence_api_id {
 
 #define DEFAULT_CODEC_ID CADENCE_CODEC_WRAPPER_ID
 
+/* we divide Cadence param id space as follows:
+ * - 0x00000000 - 0x7FFFFFFF, Cadence library ids which come
+ *   from xa_*_dec_api.h
+ * - 0x80000000 - 0xFFFFFFFF, SOF remapping of Cadence library ids
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply src/audio/module_adapter/module/cadence.c:41]: Remapping is required because there is no way we can transmit both AAC and MP3 parameters in the same param id space in the first place as they overlap. More than that they have different IDs.

For example, in order to set pcm width size we need to sent to Cadence library following parameters

MP3:

param_id:   XA_MP3DEC_CONFIG_PARAM_PCM_WDSZ  (= 0),  value = 32

AAC:

param_id:   XA_AACDEC_CONFIG_PARAM_PCM_WDSZ  (= 7), value = 32 

See: 

https://github.com/thesofproject/sof/blob/main/src/include/sof/audio/cadence/aac_dec/xa_aac_dec_api.h#L29

vs

https://github.com/thesofproject/sof/blob/main/src/include/sof/audio/cadence/mp3_dec/xa_mp3_dec_api.h#L32

File: src/audio/module_adapter/module/cadence.c (hunk 1 above)
Comment (dbaluta) [reply src/audio/module_adapter/module/cadence.c:42]: We need to store in topology parameters which basically cannot be passed via Compress API kernel-userspace interface.

E.g parameters that cannot be stored in `snd_codec` structure.

See: https://github.com/thesofproject/linux/blob/topic/sof-dev/include/uapi/sound/compress_params.h#L410

One such parameter is `PCM width size`, size of PCM output samples in bits. 

The commit message basically says: 

1) add required parameters via topology

2) enhance current implementation to dynamically setup the rest of parameters via `set_param` calback of Compress API.

For example, for AAC we could support various formats (ADTS, ADIF) etc but we currently are limited to only supporting one format (ADTS) as this is the one we could add in the topology.

In the future, I will remove this info from topology and dynamically set it by actually parsing AAC file and send it to FW.

File: tools/topology/topology1/development/sof-imx8-compr-wm8960.m4 [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+dnl     period, priority, core,
+dnl     pcm_min_rate, pcm_max_rate, pipeline_rate,
+dnl     time_domain, sched_comp)
+
+# Low Latency playback pipeline 1 on PCM 0 using max 2 channels of s32le.
+# Set 1000us deadline with priority 0 on core 0
+PIPELINE_PCM_ADD(sof/pipe-codec-adapter-playback.m4,
+	1, 0, 2, s32le,
+	1000, 0, 0,
+	48000, 48000, 48000)
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply tools/topology/topology1/development/sof-imx8-compr-wm8960.m4:66]: > @dbaluta do you have an example of the command line that would be used in this use case ? I'm guessing it's just
> 
> ```
> cplay file.mp3
> cplay file.aac
> ```
> 
> i.e. there is no kcontrol to switch between aac and mp3 ?

No need for control  to switch between aac and mp3. Compress API offers a structure `snd_codec` which encapsulates a lot of stream properties (e.g codec_id)

See: https://github.com/thesofproject/linux/blob/topic/sof-dev/include/uapi/sound/compress_params.h#L410

We use this structure dynamically at runtime to pass desired parameters to FW.

So, basically we run something like this:

cplay -I AAC   file.aac
cplay -I MP3  file.mp3

`-I` cplay could even infer the compress stream type by looking at file header. This is not implemented though.

File: tools/topology/topology1/development/sof-imx8-compr-wm8960.m4 [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+`	0x07,0x20,0x00,0x80,0x0C,0x00,0x00,0x00,'
+`	0x18,0x00,0x00,0x00,0x03,0x20,0x00,0x80,'
+`	0x0C,0x00,0x00,0x00,0x02,0x00,0x00,0x00,'
+`	0x00,0x10,0x00,0x80,0x0C,0x00,0x00,0x00,'
+`	0x18,0x00,0x00,0x00"'')
+
+define(`CA_SETUP_CONTROLBYTES_MAX', 300)
+
+undefine(`DAI_PERIODS')
+define(`DAI_PERIODS', 8)
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply tools/topology/topology1/development/sof-imx8-compr-wm8960.m4:53]: I think it might work with 2 or 3 as usual, i remember that I added this in the beginning to compensate for the non-symetric sizes of codec_adapter input and output.

Eg. the decoder will output a data much bigger at the output after decompression. I could revise this at some point. But this proved that just works. 

We have lots of memory for buffers so it was never a concern.

File: tools/topology/topology1/development/sof-imx8-compr-wm8960.m4 (hunk 6 above)
[the remark this answers is not in the document]
Comment (dbaluta) [reply tools/topology/topology1/development/sof-imx8-compr-wm8960.m4:66]: > so what happens if you decode a file encoded at e.g. 16 kHz

We only support 48Khz for now, but I think this is a limitation of SOF design. We cannot easily change backend parameters. 

As of now, most of the usecases just add a SRC in the pipeline to fix that.

**Review by lgirdwood — COMMENTED** [review]
@dbaluta think we are good to merge, I would just update the topology stream direction arrows to avoid confusion.
@plbossart good for you ?

File: tools/topology/topology1/development/sof-imx8-compr-wm8960.m4 (hunk 7 above)
[the remark this answers is not in the document]
Comment (plbossart) [reply tools/topology/topology1/development/sof-imx8-compr-wm8960.m4:53]: I am afraid this hides a scheduling problem. In some cases a frame will generate double-digit ms, and it's nearly impossible to complete the decoding for one frame in 1ms. The point is that decoding should not be a low-latency task and that there should be enough intermediate buffering between the decoder and the dai. The DAI period should not depend on what's upstream.

File: tools/topology/topology1/development/sof-imx8-compr-wm8960.m4 (hunk 2 above)
[the remark this answers is not in the document]
Comment (plbossart) [reply tools/topology/topology1/development/sof-imx8-compr-wm8960.m4:29]: change needed then

File: tools/topology/topology1/development/sof-imx8-compr-wm8960.m4 (hunk 3 above)
[the remark this answers is not in the document]
Comment (plbossart) [reply tools/topology/topology1/development/sof-imx8-compr-wm8960.m4:88]: change needed then

File: tools/topology/topology1/development/sof-imx8mp-compr-aac-wm8960.m4 (hunk 4 above)
[the remark this answers is not in the document]
Comment (plbossart) [reply tools/topology/topology1/development/sof-imx8mp-compr-aac-wm8960.m4:48]: I am still in the dark on how parameters are sent, and which ones are payload dependent and which ones aren't.

File: src/audio/module_adapter/module/cadence.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (plbossart) [reply src/audio/module_adapter/module/cadence.c:41]: not following this at all.

Is this saying that the parameters in topology need to be formatted so as to match the format expected by a specific implementation? The whole point was to try and push this sort of knowledge in the firmware, userspace and kernel shouldn't depend on what the library requires.

File: src/audio/module_adapter/module/cadence.c (hunk 1 above)
Comment (plbossart) [reply src/audio/module_adapter/module/cadence.c:42]: not following this either. how is this parameter passed to the firmware? with the set_params?

File: src/audio/module_adapter/module/cadence.c [hunk 8]
```diff
@@ -95,6 +113,46 @@ static struct cadence_api cadence_api_table[] = {
...
+
+	/* param id uses remapped value, fixup it up to original Cadence value */
+	switch (cd->api_id) {
+	case  CADENCE_CODEC_MP3_DEC_ID:
+		start = SOF_CADENCE_MP3_PARAM_ID_START;
+		break;
+	case  CADENCE_CODEC_AAC_DEC_ID:
+		start = SOF_CADENCE_AAC_PARAM_ID_START;
+		break;
+	default:
```
Comment (plbossart) [inline src/audio/module_adapter/module/cadence.c:146]: do we really want to do this with code, or should there be a table for remapping? This code would need to be modified for every codec addition, that's not so good.

**Review by plbossart — COMMENTED** [review]
Not trying to be difficult but that PR is very very hard to follow.

File: src/audio/module_adapter/module/cadence.c (hunk 8 above)
Comment (dbaluta) [reply src/audio/module_adapter/module/cadence.c:146]: @plbossart i was thinking about a table for remapping. Will try to implement it in next iteration.

File: src/audio/module_adapter/module/cadence.c (hunk 1 above)
Comment (dbaluta) [reply src/audio/module_adapter/module/cadence.c:42]: @plbossart yes, via 

```
-> ioctl(compress->fd, SNDRV_COMPRESS_SET_PARAMS, &params)
       -> `sof_compr_set_params`
       
```

**Comment by dbaluta** [comment]
Quoting @plbossart 

> Is this saying that the parameters in topology need to be formatted so as to match the format expected by a specific implementation? The whole point was to try and push this sort of knowledge in the firmware, userspace and kernel shouldn't depend on what the library requires.

Yes, unfortunately this was the initial assumption done when we introduced topology setup bytes. 

```
# Codec Adapter setup config control bytes (little endian)
#  : bytes "abi_header, ca_config, [codec_param0, codec_param1...]"
#  - 32 bytes abi_header: you could get by command "sof-ctl -t 0 -g <payload_size> -b"
#    - [0:3]: magic number 0x00464f53
#    - [4:7]: type 0
#    - [8:11]: payload size in bytes (not including abi header bytes)
#    - [12:15]: abi 3.1.0
#    - [16:31]: reserved 0s
#  - 20 bytes ca_config: codec adapter setup config parameters, for more details please refer
#                        struct ca_config under audio/codec_adapter/codec/generic.h
#    - [0]: API ID, e.g. 0x01
#    - [1:3]: codec ID, e.g. 0xd03311
#    - [4:7]: reserved 0s
#    - [8:11]: sample rate, e.g. 48000
#    - [12:15]: sample width in bits, e.g. 32
#    - [16:19]: channels, e.g. 2
# - (optional) 12+ bytes codec_param: codec TLV parameters container, for more details please refer
#                                     struct codec_param under audio/codec_adapter/codec/generic.h
#    - [0:3]: param ID
#    - [4:7]: size in bytes (ID + size + data)
#    - [8:n-1]: data[], the param data
```

I think your idea, to push the inner details in FW , kernel or userspace is good at first glance, but because we have dozens of parameters would be very hard to hide this from topology right now. 

What we can really do in the future is to set most of the parameters dynamically at runtime via an ioctl. For now, I will try to do as little damage as possible to existing implementation and move to dynamic approach in a next patch.

Actually, just right now we need a way to specify `PCM bit width size` from topology. This information is nowhere to be found in compress API parameters structs (e.g `snd_codec`) and we don't have any way to pass it at runtime. 

So, it needs to stay in topology. But to make things worse, the actual param_id for `PCM bit width size` is different from codec to codec. 

So, if we want a topology to support multiple codecs we need to have multiple entries in the topology setup bytes.

**Comment by plbossart** [comment]
@dbaluta the premise of the ioctl was to standardize on common parameters. If we abuse this ioctl and make it dependent on the fimware implementation, it's a very large deviation from the initial idea.

Also I don't get why the codec PCM width cannot be adjusted by the module interface itself.  We shouldn't manage this at the topology level, this is an implementation detail that needs to be accounted for by the module interface proper.

**Comment by lgirdwood** [comment]
> @dbaluta the premise of the ioctl was to standardize on common parameters. If we abuse this ioctl and make it dependent on the fimware implementation, it's a very large deviation from the initial idea.
> 
> Also I don't get why the codec PCM width cannot be adjusted by the module interface itself. We shouldn't manage this at the topology level, this is an implementation detail that needs to be accounted for by the module interface proper.

@plbossart are you suggesting some wrapping of the ioctl in alsa-lib with an alsa API ?? i.e. to force the standardization and help application code with config ?

**Comment by dbaluta** [comment]
> @dbaluta the premise of the ioctl was to standardize on common parameters. If we abuse this ioctl and make it dependent on the fimware implementation, it's a very large deviation from the initial idea.
> 
> Also I don't get why the codec PCM width cannot be adjusted by the module interface itself. We shouldn't manage this at the topology level, this is an implementation detail that needs to be accounted for by the module interface proper.

I wasn't suggesting to abuse the ioctl interface. 

My intention is like this:

1) - use ioctl parameters for everything it already supports. Don't add any new field. 
2) - use topology for any non-standard parameter OR not-supported-by ioctl.

Module interface cannot make decision by itself. e.g use 32bit width PCM or 24bit width PCM. It needs someone externally to instruct it.

@plbossart @lgirdwood @ranj063 do you have any 30-45 minutes next week to setup a call?

**Comment by lgirdwood** [comment]
> @plbossart @lgirdwood @ranj063 do you have any 30-45 minutes next week to setup a call?

Good for me Wednesday onwards, but @ranj063 is out.

**Comment by plbossart** [comment]
> Module interface cannot make decision by itself. e.g use 32bit width PCM or 24bit width PCM. It needs someone externally to instruct it.

Why not? don't we have information on component sink and source buffer that could be used to infer what the desired bit depth is? 

> @plbossart @lgirdwood @ranj063 do you have any 30-45 minutes next week to setup a call?

Fine with me. it wouldn't hurt to review all the known issues with the compressed support, it seems we have a few still.

**Comment by dbaluta** [comment]
@plbossart as discussed yesterday this PR is as good as it gets. 

I find @ujfalusi proposals to group parameter by codec id, e.g

[codec_id1, (param_id, value), (param_id, value)],
[codec_id2, (param_id, value), param_id, value)]

very interesting, so I will hold this PR a little bit more to get a chance to discuss with Ranjani. 

The drawback for @ujfalusi proposal is that we won't be able to keep backward compatibility between older SOF version and new topology. I'm not sure though we really want that.

**Comment by lgirdwood** [comment]
@plbossart @dbaluta sorry, I missed the call - what is the update/plan here ?

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by lgirdwood** [comment]
@dbaluta any update ?

**Comment by dbaluta** [comment]
will prepare a simpler PR for this.
